### DISCUSSION TITLE
guides/adding-custom-structures/

### DISCUSSION THREAD

**Discussion by giscus** [misode/misode.github.io#226] [Site]
# guides/adding-custom-structures/

https://misode.github.io/guides/adding-custom-structures/

**Comment by Apollounknowndev** [comment]
absolutely amazing that you're doing guides now!

**Comment by SirPankratz** [comment]
How would you set the y height for a structure?

**Reply by misode** [reply]
In 1.19 it's easy with the `start_height` field.

In 1.18.2 it's a bit tricky but it can be done by using the `bastion_remnant` structure feature type and adding an intermediate jigsaw piece to move the start. Bastion remnants always spawn at Y=33.

**Comment by ManosSef** [comment]
Will there be guides for other parts of data packs? (I mean, I won't need them, but it will be cool if they exist)

**Reply by misode** [reply]
Yes I have more guides planned, in the near future mostly focused on worldgen. Let me know if you have a specific suggestion though!

**Reply by ManosSef** [reply]
If you want, I could help with the guides for things other than worldgen, especially advancements. I love testing all kinds of different things with them and I have learned a lot.

**Reply by misode** [reply]
I'm open for PRs to contribute guides! I'll add some instructions on how to get development setup to the readme.

**Reply by AirplaneA220** [reply]
A suggestion I have is maybe how to add custom structures to the custom biomes(I mean this as like the feature placement) to kinda build off just adding new structures in the game

**Comment by xana43** [comment]
For some reason when I try this it always keeps saying "Unknown registry key in ResourceKey[minecraft:root / minecraft:worldgen/structure_feature]" I get that I need to register the structure, but how do I do that seeing as apparently that API is completely gone as of 1.18.2

**Reply by xana43** [reply]
NOTE: I'm trying to make structure generate using Fabric MC

**Reply by skerit** [reply]
I'd like to know too, getting the same error.

**Reply by misode** [reply]
If you're updating to 1.19, here is a tutorial mod for fabric: https://github.com/TelepathicGrunt/StructureTutorialMod/tree/1.19.0-Fabric-Jigsaw

**Reply by skerit** [reply]
Yup, I was still on 1.18 - I did not see the little version switcher in the top right of the article. It's working now.

**Reply by Danjhop4** [reply]
How do you add a structure that does not generate in the void in the end

**Comment by ghost** [comment]
Hi,

Brilliant work. I don't know if I would have succeeded with my plans without the things you've shared on your page.

I'd like to know if it's possible to make a structure generate in a grid, so that the structures are always an equal distance from each other. Do you know if this is possible? The 20w14infinite snapshot had it in the fleet dimension.

**Reply by misode** [reply]
You can do this by setting `separation` one less than `spacing` in the structure set.

**Reply by ghost** [reply]
I did that earlier, but it didn't work. I now retried just to be sure, and I can confirm that it didn't solve the problem. The structures are clearly placed unevenly.

Also I don't know what the "starting block" of the structure is (as in, the block that's used as the "pivot" when the structure is placed in different orientations) and what determines its coordinates inside a chunk. These certainly don't help.

**Comment by Gurkenkoenig** [comment]
Man. Just 6 months ago i found this site. I Was trying to make a village using jigsaw blocks, But i gave up. But maybe now that you write some guides i can try again. Idk how you feel about making some tutorials on utube. Thanks for your time.

**Comment by GentlemanRevvnar** [comment]
Are there any  other variations beside
**"terrain_adaptation": "beard_thin"**   ?

**Reply by misode** [reply]
Thanks for reminding me, I was supposed to list these options in the guide. For you can see the options in the generator.

**Comment by ThatRobin** [comment]
Im currently working on a structure, but each time it generates, it generates 1 block into the ground, is there a way to stop it from doing that? All of my setting match the tutorails exactly, im just using using my own structure

**Reply by misode** [reply]
If you're in 1.19 you can set `"start_height": { "absolute": 1 }`, if you're in 1.18.2 you just have to add an extra layer of structure voids at the bottom.

**Comment by Silabear** [comment]
Why hello there,

Is there any way to make these spawn in a nether-like dimension, since it seems to spawn on the roof for me. Thanks :D

**Reply by misode** [reply]
It's impossible right now to make a structure spawn on a surface that is not the very top surface of the world (cave or nether like dimension). If you notice both the bastion remnant and ancient city spawn at a height independent of the terrain.

**Reply by Silabear** [reply]
So essentially I can't?

**Comment by mbg206** [comment]
Where did you find this information? I am trying to figure out what each of the options for "terrain_adaptation" and "project_start_to_heightmap" do, but I can't seem to find any documentation online.

**Reply by misode** [reply]
Most of this comes from looking at the vanilla worldgen files, and from reading decompiled source code. I've just added the info for `terrain_adaptation`.

**Comment by mbg206** [comment]
What's the difference between "WORLD_SURFACE" and "WORLD_SURFACE_WG", and what exactly is "MOTION_BLOCKING" and "MOTION_BLOCKING_NO_LEAVES"?

**Reply by misode** [reply]
The different types are explained in a [separate guide](https://misode.github.io/guides/heightmap-types/)

**Comment by Lonemind** [comment]
How would one go about making half buried structures in 1.18.2?

**Reply by Lonemind** [reply]
Misode was kind enough to reply on his discord. For those looking here for answers the solution in 1.18.2 is to split your structure into two pieces, a top and a bottom and then link them with jigsaws.

Apparently, though, 1.19 has the "start_height" option which will make this easy with just a single structure.

**Comment by Gurkenkoenig** [comment]
would it be posible to make the tower guide but with a jigsawblock? i cant really get it to work, i am trying to look at other datapacks online that do stuff with jigsaw blocks but i cant figure it out.

**Reply by misode** [reply]
There's this [very detailed guide for jigsaw structures](https://gist.github.com/GentlemanRevvnar/387f9ee28613715c187a36dbc1dff35d) by GentlemanRevvnar

**Reply by Gurkenkoenig** [reply]
> 

Cheers man. I somehow didnt get an Email 11 days ago. I check it out when i am home.

**Comment by BluJayy728** [comment]
I've got structures generating in the world fine, but for some reason they are all sunk 1 block into the ground. Why is it doing this, and how do I fix it? I'm on 1.19 if it matters.

**Reply by Gurkenkoenig** [reply]
Is it allowed to link me your Pack? Did you use the wg generation step Method or the non wg thingy? I am at work so i cant check what the names are

**Reply by BluJayy728** [reply]
I'm using WORLD_SURFACE_WG. I'm thinking I might just need to put structure void under my structures.

**Reply by BluJayy728** [reply]
Yeah all I had to do was use structure voids under the structure. Thanks for trying to help though.

**Comment by TimmyChips** [comment]
What would be the best way to generate structures that use the `underground_structures` tag? Using `WORLD_SURFACE_WG` for `project_start_to_heightmap` doesn't seem to always generate the structures. Removing the heightmap makes it so structures will actually spawn, except sometimes when they spawn, they're by themselves and are not connected to any caves. There might not be any work around though, it might just be like dungeons, which a lot of times spawn not connected to anything.

**Reply by TimmyChips** [reply]
Ideally I would like to make structures that are on the sides of cave walls similarly to dungeons and amethyst geodes, and spawn closer to the bottom of the world.

**Reply by misode** [reply]
There's currently no way to get an underground structure to spawn like that unfortunately

**Reply by TimmyChips** [reply]
Ah gotcha, thank you for the response! I figured there wouldn't be a way to do it, but I figured I would ask. I guess you could do `thin_beard` to try to open it more, or a workaround would be to create several "cave carver" structures and use jigsaws to try to branch out.

**Comment by minecraftmaker13** [comment]
can it spawn in other dimensions such as the end?

**Comment by Bret06** [comment]
These guides are cracked✨

**Comment by randomgingerguy** [comment]
Do you know of any processors or ways to avoid vanilla generation destroying underground structures? Azalea trees and magma blocks are replacing my structure, fossils are appearing inside of the structure, and waterloggable blocks are being flooded, and I'm not sure how to fix this.
I only use structure void outside of the structure shell, I am using 'underground structures' step, 'rigid' projection, and 'beardbox' element type.  Any help would be immensely appreciated.

**Comment by rubmurfer** [comment]
Is there a way to make the structure generate between two variables? example: y=-20 to y=20

**Reply by randomgingerguy** [reply]
You would use the 'weighted towards bottom' option, and set the y-heights you want.  Then, it'll generate anywhere between them.  That's what I did, at least.

**Comment by InteJason** [comment]
Do structures rotate when they generate? If so how can you stop that from happening?

**Reply by Gurkenkoenig** [reply]
What do you mean with rotate? Do you mean the starting structure?

**Reply by InteJason** [reply]
Lets say I want my structure's entrance to face north. When it actually generates in-game, will it rotate the structure so that the entrance faces other direction instead? If so, what can I do to fix it?

**Reply by Gurkenkoenig** [reply]
Idk But you can test it easily by just building a xyz cross with Red blue green wool like in the f3 Mode and test it. I am to lazy right now.

**Comment by Epicman44553** [comment]
My structure is located in the sky, and when I load into my test world there's a dirt island under it anybody know how to fix this?

**Reply by Gurkenkoenig** [reply]
Change beardthin to whatever doesnt generate land

**Reply by Epicman44553** [reply]
ok

**Comment by MTGPROD** [comment]
Your guide is amazing, If you could dig deeper (especially on how to make custom villages) that would be incredible. Thank you !

**Reply by Gurkenkoenig** [reply]
https://gist.github.com/GentlemanRevvnar/98a8f191f46d28f63592672022c41497

I am sure you can use this for great villages. Its better than misodes guide in my opinion. Because it has more infos

**Reply by MTGPROD** [reply]
Thank you, that's what I searched

**Comment by Netwyn** [comment]
Is there a way we can set random things inside of the structure? For example, like how not all coral reefs are the same; some have different types of coral blocks, and those coral blocks are all spread/spaced differently.

**Reply by misode** [reply]
Coral reefs are features and generated with code. You can get some randomization with jigsaws in structures, but not nearly enough to replicate corals

**Reply by DeLoR1AnEc** [reply]
U can use more pools in template pool

**Comment by IMB11** [comment]
Is there any way to set a max y height for the structure? Similarly to start y?

**Comment by andrbrow** [comment]
Love the Guide.

Question about the Salt parameter since I've seen mods that have the same salt value between different structures contrary to your advise here.  I'm trying to spread out some structures on their mods for my own gameplay but running into issues.  Thanks!

**Comment by Matthgeek** [comment]
How would you replace a structure? Say you want to change out dungeons for something more exciting.

**Reply by DeLoR1AnEc** [reply]
create the same name file in data/minecraft/...

**Comment by 1010allen1010** [comment]
how do i use the offset functionality of the "start_height" field,

it does not seem to do anything using the "WORLD_SURFACE" or "WORLD_SURFACE_WG"
i need my structure 3 block in the ground

**Comment by whiderboss** [comment]
How can I prevent a structure from generating in water?

**Comment by YeahBoi321** [comment]
How would you get a jigsaw structure to generate instead of just a static structure?

**Reply by GentlemanRevvnar** [reply]
Maybe this guide will help somewhat
https://gist.github.com/GentlemanRevvnar/98a8f191f46d28f63592672022c41497

**Comment by crosscountrycoder** [comment]
How can I add a custom structure (or a built-in Minecraft structure) at a specific location? For example, if I want to create a stronghold at (-1239, 17, -315), which an eye of ender will then point to, how do I do that?

**Reply by crosscountrycoder** [reply]
Edit: add a custom structure in an already existing world

**Comment by Luke652** [comment]
What's the size limit for a Structure now? I know that 1.18 had a limit of 7, but what does 1.19 have?

**Reply by misode** [reply]
It has the same jigsaw depth limit of 7. In terms of blocks there is a limit of 128 blocks out from the center.

**Comment by Alexey10240** [comment]
Will there be a tutorial on creating processor lists?

**Comment by Insane96** [comment]
There's anyway in 1.18.2 to prevent the structure from generating inside an area (E.g. 200 blocks from 0, 0)?

**Reply by misode** [reply]
Not really. You could pregen the world around 0 0 with a version of the data pack that doesn't have the structure if that works for you.

**Reply by TheeMrCheeky** [reply]
I set it to underground but still spawns ontop on the bedrock in nether ugh idk really trying to figure this out. ty in advance

**Comment by TheeMrCheeky** [comment]
how would I make a structure spawn in nether?

**Reply by TheeMrCheeky** [reply]
by this i mean they spawn but they spawn up on bedrock

**Reply by TheeMrCheeky** [reply]
I set it to underground but still spawns ontop on the bedrock in nether ugh

**Reply by fishBone000** [reply]
i guess it's related to the project_start_to_heightmap. Check vanilla nether structure files as examples.

**Reply by Alumopper** [reply]
just set "start_height" like this
```
"start_height": {
      "absolute": -70
    }
```
change the value of "absolute" to change the depth of your structures from the top.

**Comment by Youritch** [comment]
hi, how can I prevent my structure to turn (if the structure is not generate in the good angle it's break it) ??

**Comment by Aitbead** [comment]
I am copy pasting original minecraft vanilla villages and editing them to spawn in the Nether. I managed to do that, but I have some problems with the natural spawning of them. In the "name_datapack\data\minecraft\worldgen\structure" folder I can edit stuff like you mentioned above, but the villages either spawn on top of the nether (on the bedrock layer) or they spawn down within the nether, but only the city centers. Do you know a way to bypass both these problems?

**Comment by ThatHomelessGuy** [comment]
I'm in 1.19 and cannot for the life of me get any structures to spawn. My structure doesn't even show up under the locate command.

**Reply by vico93** [reply]
That's why i absolutely hate the vanilla standard about this (and worldgen in general as well). While misode did a awesome work documenting that i still can't understand well and would kill for someone to bring Pillar back to life, this time as an alternative specification for custom structure spawning.
https://legacy.curseforge.com/minecraft/mc-mods/pillar
https://docs.google.com/document/d/13wnL56VXB76Yth3lY__hP_8WAu5eYNWls0q4N-3VpVo/edit
https://github.com/VazkiiModsArchive/Pillar

**Comment by PansRB** [comment]
does someone know how do i make the structure and turn created structure into nbt thing?

**Reply by Insane96** [reply]
Structure blocks https://minecraft.fandom.com/wiki/Structure_Block

**Reply by stefann01** [reply]
This guy [here](https://www.youtube.com/watch?v=Ksli-clsBnk&t=53s) explains.

**Comment by vico93** [comment]
Are custom structures like this ones generate in overworlds of custom world presets (in 1.20)?

**Comment by majster2nn** [comment]
So yeah the datapack is created one small issues: structures dont spawn and minecraft seem not to detect that they exist

**Reply by DaNitroNinja** [reply]
Yh same with me. I used 1.20 and tried 1.19

**Reply by misode** [reply]
Could you share your data pack here so someone can take a look and check what the issue is?

**Reply by DaNitroNinja** [reply]
Should I just send the text and the directory?

**Reply by DaNitroNinja** [reply]
data/impossible/structures/jungle_home.nbt (I cant share the nbt, but I checked and if i use a structure block this will appear)

data/impossible/worldgen/jungle_home_set/jungle_home.json:

{
    "structures": [
      {
        "structure": "impossible:structures/jungle_home",
        "weight": 1
      }
    ],
    "placement": {
      "type": "minecraft:random_spread",
      "spacing": 5,
      "separation": 2,
      "salt": 1646207470
    }
  }

data/impossible/worldgen/jungle_structure/jungle_home.json:

{
    "type": "minecraft:jigsaw",
    "biomes": "minecraft:plains",
    "step": "surface_structures",
    "spawn_overrides": {},
    "terrain_adaptation": "beard_thin",
    "start_pool": "impossible:template_pool/jungle_home",
    "size": 1,
    "start_height": {
      "absolute": 0
    },
    "project_start_to_heightmap": "WORLD_SURFACE_WG",
    "max_distance_from_center": 80,
    "use_expansion_hack": false
  }

data/impossible/worldgen/template_pool/jungle_home.json:

{
    "name": "impossible:jungle_home",
    "fallback": "minecraft:empty",
    "elements": [
      {
        "weight": 1,
        "element": {
          "element_type": "minecraft:single_pool_element",
          "location": "impossible:structures/jungle_home",
          "projection": "rigid",
          "processors": "minecraft:empty"
        }
      }
    ]
  }

**Reply by DaNitroNinja** [reply]
This is all of the code from the structures and worldgen folder in my datapack. The name space is impossible.

**Reply by DaNitroNinja** [reply]
Sorry, github on my computer is saying that the messages above didn't send, please can you confirm whether they have sent now?

**Reply by misode** [reply]
The folder of one of your files is incorrect. Move `data/impossible/worldgen/jungle_structure/jungle_home.json` to `data/impossible/worldgen/structure/jungle_home.json`.

And also replace the reference in the structure set to: `"structure": "impossible:jungle_home",`

**Reply by DaNitroNinja** [reply]
Thank you, minecraft now says it exists, but if i use '/locate' it will just give an error message saying "cannot find a structure of type impossible:jungle_home nearby.

**Reply by DaNitroNinja** [reply]
I managed to fix it!. I forgot to call the structure_set structure_set. That was stupid of me!

**Comment by DaNitroNinja** [comment]
For me if i try to use the datapack, it says that it fails validation, even on a new world.

**Reply by JjCraft909** [reply]
same

**Comment by ShdwDrgnVIIv2** [comment]
Could someone have a look at my pack and see why my structure_set files are giving me an error with the namespace? It would be much appreciated!

https://github.com/ShdwDrgnVIIv2/test_pack/tree/main/data

**Comment by JjCraft909** [comment]
I used this in my datapack and it no longer works after I added a second Structure. It shows only one error in the structure_set files: "Failed to resolve namespaced ID “sqol:undead_temple” in cache category “worldgen/configured_structure_feature”.datapack(4)". I am creating a 1.20 Datapack and would like to know what changed from 1.19 to 1.20 that causes this bug? idk how to send my code but if i did I would do so.

**Comment by codemdragon** [comment]
How do I add this to a specific dimension

**Comment by Julynx** [comment]
I'm making a "sinkhole" structure. 

The "/locate structure" command autocompletes the name, but when I hit enter it just always says "Could not find a structure of type sinkhole:sinkhole nearby"

The "/place structure" command just says "Failed to place structure" every single time.

I want the sinkhole to be buried in the terrain, with only the top side open so that the hole is open from above.
